### PR TITLE
Add deleted_on field to AutomaticUpdateRule, Broadcast, and Schedule

### PR DESCRIPTION
**Pull request by gherceg**
## Technical Summary

https://dimagi-dev.atlassian.net/browse/SAAS-14571

Adds `deleted_on` field to the following SQL tables:

- AutomaticUpdateRule
- Broadcast
- Schedule

Additional PRs will be created that migrate remaining tables.

The `deleted_on` field will accomplish what the `deleted` field currently does (informing us whether an object is deleted or not), while also giving valuable information of _when_ the object was deleted. This will be used by a periodic task that permanently deletes soft deleted objects older than 30 days.

In addition to adding a `deleted_on` field, this PR does two important things.

#### Partial Index
The partial index on the `deleted_on` column will enable faster queries by the periodic task when searching for objects deleted over 30 days ago.

#### Set `deleted_on` for existing objects
Rather than handle this in the periodic task, it makes sense to get this out of the way. The `deleted` field can be removed once this PR has been merged and deployed (aka the migration has been run). The migration code ensures that the `deleted_on` field is set to the current time for objects with `deleted` or `is_deleted` already set to True.

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code
The migration includes setting `deleted_on` for existing soft deleted objects that are null. This PR needs to also include setting `deleted_on` whenever the object is deleted to catch any objects that are soft deleted _after_ the migration has been run. If it goes out separately, we run the risk of a gap in setting `deleted_on` on objects that have been deleted.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by gherceg** [comment]
Starting as a draft to make sure the path I'm on so far is aligned with reviewers, mainly @biyeun and @millerdev.

File: corehq/apps/data_interfaces/migrations/0033_automaticupdaterule_deleted_on.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        ('data_interfaces', '0032_bootstrap_audit_events_for_update_rules'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='automaticupdaterule',
+            name='deleted_on',
+            field=models.DateTimeField(null=True),
+        ),
+        migrations.RunSQL(CREATE_INDEX),
```
Comment (millerdev) [inline corehq/apps/data_interfaces/migrations/0033_automaticupdaterule_deleted_on.py:29]: It's possible to create partial indexes in Django models, which are then automatically detected by the migration framework. Example:

https://github.com/dimagi/commcare-hq/blob/5c59cd06fa808515f3d46ba1282767eb8ee801f2/corehq/blobs/models.py#L77-L81

File: corehq/messaging/scheduling/migrations/0025_schedule_deleted_on.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            name='deleted_on',
+            field=models.DateTimeField(null=True),
+        ),
+        migrations.AddField(
+            model_name='timedschedule',
+            name='deleted_on',
+            field=models.DateTimeField(null=True),
+        ),
+        migrations.RunSQL(ALERT_CREATE_INDEX),
+        migrations.RunSQL(TIMED_CREATE_INDEX),
```
Comment (millerdev) [inline corehq/messaging/scheduling/migrations/0025_schedule_deleted_on.py:39]: Suggestion: add partial indexes in Django model.

**Review by millerdev — COMMENTED** [review]
Looking good.

File: corehq/apps/data_interfaces/models.py [hunk 3]
```diff
@@ -134,6 +134,11 @@ class CriteriaOperator(models.TextChoices):
 
     class Meta(object):
         app_label = "data_interfaces"
+        indexes = [
+            models.Index(fields=['deleted_on'],
+                         name='rule_deleted_on_idx',
```
[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/apps/data_interfaces/models.py:139]: django has a limit of 30 characters on the index name which made me shorten it. If you have any suggestions on a format to fit all of that in 30 characters, I'm open to it!

File: corehq/messaging/scheduling/migrations/0025_schedule_deleted_on.py (hunk 2 above)
Comment (gherceg) [reply corehq/messaging/scheduling/migrations/0025_schedule_deleted_on.py]: d36834a865b650d30d5bad35a814c18557e1af23

File: corehq/apps/data_interfaces/migrations/0033_automaticupdaterule_deleted_on.py (hunk 1 above)
Comment (gherceg) [reply corehq/apps/data_interfaces/migrations/0033_automaticupdaterule_deleted_on.py]: d36834a865b650d30d5bad35a814c18557e1af23

**Review by snopoke — COMMENTED** [review]
Would it be useful to have a 'deleted_by' field? Or is that info already captured in the audit logs?

**Comment by gherceg** [comment]
> Would it be useful to have a 'deleted_by' field? Or is that info already captured in the audit logs?

I can't think of a reason to have the `deleted_by` field for what I'm working on specifically. That info would be captured for the `AutomaticUpdateRule` since that is being audited, but not for other models.

File: corehq/apps/data_interfaces/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/apps/data_interfaces/models.py]: See bd8dca88e4e0b5c26882fab43bda59a3c255e224 and dfa19f7c8dc7670ca4fbbe79d449d3f617d89383

**Review by millerdev — APPROVED** [review]
Approving since this seems like it will work, although I think it would be nice to use the same logic that Django uses to generate the name. See [this comment](https://code.djangoproject.com/ticket/23577#comment:24) for an example of how to do that.

**Comment by gherceg** [comment]
> although I think it would be nice to use the same logic that Django uses to generate the name. See [this comment](https://code.djangoproject.com/ticket/23577#comment:24) for an example of how to do that.

I looked into this but hit multiple issues. The first being that the built-in method doesn't offer a way to limit the index name to 30 characters, so Django still complains and we would need to manually intervene. The second is that it is a challenge to reference the `deleted_on` field from inside of the Meta class object, which could be worked around by passing in hard coded strings of app name and field, but by then it is starting to look pretty similar to the current implementation.

The main difference is the current implementation appends `_idx` to the end of the name, whereas Django ends with the hash. This is actually good as it further reduces the likelihood of any index name conflicts.

All that to say, we are sticking with the current implementation 👍🏻

**Comment by gherceg** [comment]
Noting that I tested this migration on staging and it didn't take long to run. Comparing number of `AutomaticUpdateRule` `Schedule` and `Broadcast` objects between staging and production, everything is below 15,000 records, some only on the magnitude of hundreds. I don't anticipate any need to run this migration separate from the production deploy.

**Pull request merged**